src/Avalonia.Visuals/Animation/Transitions/Rotate3DTransition.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Avalonia.Media;
+using Avalonia.Styling;
+
+namespace Avalonia.Animation;
+
+public class Rotate3DTransition: PageSlide
+{
+    
+    /// <summary>
+    ///  Creates a new instance if the <see cref="Rotate3DTransition"/>
+    /// </summary>
+    /// <param name="duration">How long the rotation should take place</param>
+    /// <param name="orientation">The orientation of the rotation</param>
+    public Rotate3DTransition(TimeSpan duration, SlideAxis orientation = SlideAxis.Horizontal) 
+        : base(duration, orientation)
+    {}
+
+    /// <summary>
+    ///  Creates a new instance if the <see cref="Rotate3DTransition"/>
+    /// </summary>
+    public Rotate3DTransition() { }
+
+    /// <inheritdoc />
+    public override async Task Start(Visual? @from, Visual? to, bool forward, CancellationToken cancellationToken)
+    {
+        if (cancellationToken.IsCancellationRequested)
+        {
+            return;
+        }
+
+        var tasks = new List<Task>();
+        var parent = GetVisualParent(from, to);
+        var (rotateProperty, center) = Orientation switch
+        {
+            SlideAxis.Vertical => (Rotate3DTransform.AngleXProperty, parent.Bounds.Height),
+            SlideAxis.Horizontal => (Rotate3DTransform.AngleYProperty, parent.Bounds.Width),
+            _ => throw new ArgumentOutOfRangeException()
+        };
+
+        var depthSetter = new Setter {Property = Rotate3DTransform.DepthProperty, Value = center};
+        var centerZSetter = new Setter {Property = Rotate3DTransform.CenterZProperty, Value = -center / 2};
+
+        if (from != null)
+        {
+            var animation = new Animation
+            {
+                Duration = Duration,
+                Children =
+                {
+                    new KeyFrame
+                    {
+                        Setters =
+                        {
+                            new Setter { Property = rotateProperty, Value = 0d },
+                            new Setter { Property = Visual.ZIndexProperty, Value = 2 },
```

Comment:
Do we really need to set ZIndex here? Do the views overlap?

Reply:
Yes, at least for a very brief moment. Having a perspective means, that the sides of a cube will always point "inwards". So we have to define a Z Index because we can't relay on the order within the XAML:

![cube](https://user-images.githubusercontent.com/374264/151357096-f86cafcb-b8d1-46bb-b348-6df22f3487a4.png)